### PR TITLE
[SCHEMA] Reorganize schema code into a package

### PR DESCRIPTION
**Pull request by tsalo**
This stems from https://github.com/bids-standard/bids-specification/pull/885#issuecomment-934535138. The idea is to make the schema code operate as a proper Python package, so we can also start writing tests for it.

We still may want to move schema-related code into a different repository at some point (e.g., `bids-schema`, `pybids`), but for now I'd just like to have the code organized nicely within this repo.

Changes proposed:

- Reorganize `schemacode` into a Python package, so that it can be installed and tests can be written for it.
    - This involves symlinking the schema files in `src/schema` to `tools/schemacode/schemacode/data/schema`. Thanks to @effigies for that.
- Write a few initial tests for `schemacode`. These can be improved in future PRs.
- Add a GitHub Action to run continuous integration for the new package and upload a coverage report to CodeCov.
- Add new schema-related folders to the CODEOWNERS file with myself listed.

### CODE REVIEW

**Comment by effigies** [comment]
@tsalo Made several updates. RTD is fixed. The PDF build looks like it needs updating to new modules. I assume you'll be able to figure that out faster than me, but let me know if you want me to tackle it.

**Comment by tsalo** [comment]
Thanks @effigies! How does it work? It looks like the schema files are duplicated in `tools/schemacode/schemacode/data/schema/`. Are those files updated automatically when the files in `src/schema/` are updated?

**Comment by effigies** [comment]
It's just a symlink, so they will be copied into the package. If you're going to develop, it probably makes sense to `pip install -e` so you don't freeze them into a package and keep using the symlink. But from a CI build perspective, it's all based on the current commit, so fine to freeze.

**Comment by tsalo** [comment]
Awesome! I fixed the module paths, so I think everything is good to go.

**Comment by tsalo** [comment]
I'd like to have a CI job that runs the package's tests (at least with Python 3.7, if not other supported versions). Is there any preference for CircleCI vs. GitHub Action?

**Comment by effigies** [comment]
GHA.

**Comment by tsalo** [comment]
Okay, I think it's looking good. I've added a few tests and a GitHub Action to run said tests. The tests are passing and the coverage report is uploading to CodeCov.

In a perfect world, I think we'd only run the tests when files in `src/schema/` or `tools/schemacode/` have been edited, but I don't want to hold this up on that basis. We could always try it in a future PR.

File: tools/schemacode/versioneer.py
Comment (sappelhoff) [inline tools/schemacode/versioneer.py:1]: why not use version 0.21?

https://github.com/python-versioneer/python-versioneer/tags

File: requirements.txt
Comment (sappelhoff) [inline requirements.txt:9]: [process_markdowns.py](https://github.com/bids-standard/bids-specification/blob/master/pdf_build_src/process_markdowns.py) depends on numpy, and depends on it being in requirements.txt --> https://github.com/bids-standard/bids-specification/blob/89115cd9ef0d4676ae78cfa1e9e78fa1048869a2/.circleci/config.yml#L66

It's also mentioned in the docs: https://github.com/bids-standard/bids-specification/blob/master/pdf_build_src/README.md

I don't know why the job still seems to pass :shrug: 

Could you either:
1. remove the numpy dependency in process_markdowns.py (preferred, seems possible at first glance, maybe even easy)
2. update the documentation (pdf build README) and ensure that numpy is installed whenever someone wants to build the pdf?
3. any other solution you might have,

**Review by sappelhoff — COMMENTED** [review]
the linkchecker issue seems to have to do with the glossary PR we recently merged. Is that an easy one to fix here as well @tsalo? Or separate PR?

File: tools/schemacode/versioneer.py
Comment (tsalo) [reply tools/schemacode/versioneer.py]: Good idea. I updated it in 31bcd196b601481b73c5c6f1101d9f60cdb37fc9.

**Comment by tsalo** [comment]
>the linkchecker issue seems to have to do with the glossary PR we recently merged. Is that an easy one to fix here as well @tsalo? Or separate PR?

@sappelhoff I think it's just because there's a "canonical" link on all pages that goes to the stable version based on the MkDocs config file (which says that the `site_url` is the stable version). Until we have a new release with the glossary page included, I think that linkchecker is going to fail.

File: requirements.txt
Comment (tsalo) [reply requirements.txt:9]: It looks like the `requirements.txt` file includes `schemacode`, which, in turn, includes `numpy` as a dependency. That explains why it's working. However, I agree that `numpy` probably shouldn't have been removed from the main requirements file, since it's used for specification code outside of `schemacode`. I'll add it back in.

The other requirements that were removed from that file all seem to be `schemacode`-specific, so I'll keep them out.

Comment (tsalo) [reply requirements.txt]: I added it back in in f766804.

Comment (sappelhoff) [reply requirements.txt]: > It looks like the requirements.txt file includes schemacode, which, in turn, includes numpy as a dependency. That explains why it's working.

thanks!

> I'll add it back in.

:+1: 

> The other requirements that were removed from that file all seem to be schemacode-specific, so I'll keep them out.

excellent, thanks

**Comment by sappelhoff** [comment]
> @sappelhoff I think it's just because there's a "canonical" link on all pages that goes to the stable version based on the MkDocs config file (which says that the site_url is the stable version). Until we have a new release with the glossary page included, I think that linkchecker is going to fail.

Mh, perhaps this could be solved by changing the `site_url` value as part of the release protocol. During release (and the version bumps) we change `site_url` to stable ... then after release (when changing the version to the dev version), we change it to "dev"

would that work?
https://github.com/bids-standard/bids-specification/blob/89115cd9ef0d4676ae78cfa1e9e78fa1048869a2/mkdocs.yml#L2

**Comment by tsalo** [comment]
I think so, but I don't know if it's worth the effort. I don't think the links are a real problem outside of linkchecker. When you click links to other pages in the specification, they go to the same version you started from, rather than defaulting to `stable`. I believe that the messed up links only appear in HTML comments in the built site.

**Comment by tsalo** [comment]
@bendhouseart sorry for not requesting your review earlier. I know you wanted to take a look, though we didn't end up trying out `poetry`.

**Review by sappelhoff — APPROVED** [review]
> I believe that the messed up links only appear in HTML comments in the built site.

I see :| but it's still annoying to have linkchecker now fail until we release ...

But that's not a blocker for this PR, which I approve of!

File: tools/schemacode/setup.cfg
Comment (yarikoptic) [inline tools/schemacode/setup.cfg:18]: what about
```suggestion
    Programming Language :: Python :: 3.9
    Programming Language :: Python :: 3.10
```
?
versioneer 0.21 is compatible according to changes done since 0.21 (just addition of 3.10 marker)

File: .github/workflows/schemacode_ci.yml
Comment (yarikoptic) [inline .github/workflows/schemacode_ci.yml:34]: what about
```suggestion
            python-version: ["3.8", "3.10"]
```
to test against "target" and the "most recent"?  or even better
```suggestion
            python-version: ["3.6", "3.7", "3.8", "3.9", "3.10"]
```
against all supported? (if adding 3.10)

File: .github/workflows/schemacode_ci.yml
Comment (yarikoptic) [inline .github/workflows/schemacode_ci.yml:5]:
```suggestion
    paths:
      - 'tools/schemacode/**'
      - 'src/schema/**'
    branches:
```
and similar one for `pull_request` to not bother testing the package (if takes long, didn't check) if no package related changes?

**Review by yarikoptic — COMMENTED** [review]
just some minor comments/suggestions - ok to ignore

File: .github/workflows/schemacode_ci.yml
Comment (tsalo) [reply .github/workflows/schemacode_ci.yml:5]: Ohhh... I didn't know that was an option! The tests are currently pretty fast, but I was worried about spamming PRs with a bunch of unnecessary checks. This should deal with that nicely.

**Comment by sappelhoff** [comment]
>>    I believe that the messed up links only appear in HTML comments in the built site.

> I see :| but it's still annoying to have linkchecker now fail until we release ...

We could add a small Python script that checks whether the current commit is tagged (=stable), or not (=dev), and then in CI **before** building with mkdocs, that Python scripts changes `stable` to `latest` in the following line:

https://github.com/bids-standard/bids-specification/blob/98ca3fa8795a2ad29910cc54327c3cedb8f84906/mkdocs.yml#L2

:thinking: or would we screw up search engine results with that? I am not very familiar with "canonical URL" metadata in HTML headers.

**Comment by tsalo** [comment]
>We could add a small Python script that checks whether the current commit is tagged (=stable), or not (=dev), and then in CI before building with mkdocs, that Python scripts changes stable to latest

I have no clue if that's possible, but having the ability to update content files with Python scripts _within_ CI would be really useful for adding tooltips (#954). The problem I have been bumping into for that is that the macros-generated content is not available for the auto-linked tooltips to grab onto. Generating the glossary within CI would circumvent that issue, I believe.

That said, I don't even know where to start to do it.

**Review by bendhouseart — APPROVED** [review]
Looked over the changed files and managed to build and install locally w/ setup.py 👍. I'd call it good to go. 

```python
Python 3.9.9 (main, Nov 21 2021, 03:22:47)
[Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import schemacode
>>>
```

**Comment by bendhouseart** [comment]
CircleCI disagrees, but eh.

File: tools/schemacode/schemacode/render.py
Comment (effigies) [inline tools/schemacode/schemacode/render.py:382]: I think you want `r"\"` or `"\\"`, but not both.

```suggestion
        suffix_str = re.findall(r"\((.+)\)", string)
```

Unless you're specifically trying to match two `\` characters?

**Review by effigies — APPROVED** [review]
Looks good. A couple bits of code caught my eye as likely able to be simplified. I haven't tested though...

**Comment by tsalo** [comment]
The built site looks good to me, as does the PDF. Now that we have two approvals, is this alright to merge or do we want to wait 5 days?

**Comment by effigies** [comment]
The 5 days is more for semantic changes than infrastructure.

**Pull request merged**
